### PR TITLE
ARM64: thread handling optimisations

### PR DESCRIPTION
**Pull request by npitre**
- arm64: improved arch_switch() implementation
- arm64: simplify user mode transition code

### CODE REVIEW

File: arch/arm64/core/switch.S [hunk 1]
```diff
@@ -40,9 +42,9 @@ SECTION_FUNC(TEXT, z_arm64_context_switch)
 	stp	x27, x28, [x1, #_thread_offset_to_callee_saved_x27_x28]
 	stp	x29, x4,  [x1, #_thread_offset_to_callee_saved_x29_sp_el0]
 
-	/* Save the current SP_ELx */
+	/* Save the current SP_ELx and return address */
 	mov	x4, sp
-	str	x4, [x1, #_thread_offset_to_callee_saved_sp_elx]
+	stp	x4, lr, [x1, #_thread_offset_to_callee_saved_sp_elx_lr]
```
Comment (carlocaione) [inline arch/arm64/core/switch.S:47]: nit: can we keep using `x30` instead of `lr` to be consistent with the current code?

**Review by andyross — APPROVED** [review]
Nice.  Yeah, that pattern[1] really never made much sense to me.  At best it saves a little bit of code size by unifying the interrupt entry/exit and context switch paths, but as this PR shows those are already unifiable with just a function call here or there.

What does this do for stability?  I've been noticing that the arm64 qemu devices tend to be a little flaky under host-stress situations, to the tune of 2-3 failures per full twister run on my system.  And I've been suspicious, but never investigated, that this was due to the need for CONFIG_SWAP_NONATOMIC[2], which this implemention cleanly eliminates.

+1 for the design choice, I'll let the ARM folks nitpick on the implementation details.

[1] Decended from the PendSV implementation for our original ARM code -- care to fix that one too?

[2] It would depend on the priority of the context switch SVC.  If it can itself be preempted, it's possible for thread mode to to mask interrupts, latch the SVC, and then unmask them only for the system to start executing a higher priority interrupt which would then see an inconsistent scheduler state (_current had been changed, but the context switch hasn't happened).

**Comment by npitre** [comment]
On Tue, 15 Mar 2022, Carlo Caione wrote:

> nit: can we keep using `x30` instead of `lr` to be consistent with the current code?

Sure. Maybe in a followup patch? Otherwise collected approvals are lost.
(This new policy is a real bummer)

Alternatively, that followup patch could turn all x30 into lr for better 
clarity?

**Comment by carlocaione** [comment]
> Alternatively, that followup patch could turn all x30 into lr for better clarity?

I think this is a great idea TBH. ACK on this one.

**Review by carlocaione — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Pull request merged**